cpp/src/experimental/louvain.cuh

```diff
@@ -671,549 +305,107 @@ class Louvain {
-    //
-    //  FIXME:  Think about how this should work.
-    //          I think Leiden is broken.  I don't think that the code we have
-    //          actually does anything.  For now I'm going to ignore Leiden in
-    //          MNMG, we can reconsider this later.
-    //
-    //  If we ignore Leiden, I'd like to think about whether the reduction
-    //  should occur now...
-    //
-
-    //
-    // Sort the results, pick the largest version
-    //
-    thrust::sort(rmm::exec_policy(stream_)->on(stream_),
-                 thrust::make_zip_iterator(thrust::make_tuple(
-                   final_src_v.begin(), final_nbr_weights_v.begin(), final_nbr_cluster_v.begin())),
-                 thrust::make_zip_iterator(thrust::make_tuple(
-                   final_src_v.end(), final_nbr_weights_v.end(), final_nbr_cluster_v.begin())),
-                 [] __device__(auto left, auto right) {
-                   if (thrust::get<0>(left) < thrust::get<0>(right)) return true;
-                   if (thrust::get<0>(left) > thrust::get<0>(right)) return false;
-                   if (thrust::get<1>(left) > thrust::get<1>(right)) return true;
-                   if (thrust::get<1>(left) < thrust::get<1>(right)) return false;
-                   return (thrust::get<2>(left) < thrust::get<2>(right));
-                 });
-
-    //
-    //  Now that we're sorted (ascending), the last entry for each src value is the largest.
-    //
-    local_cluster_edge_ids_v.resize(final_src_v.size());
-
-    thrust::transform(rmm::exec_policy(stream_)->on(stream_),
-                      thrust::make_counting_iterator<edge_t>(0),
-                      thrust::make_counting_iterator<edge_t>(final_src_v.size()),
-                      local_cluster_edge_ids_v.begin(),
-                      [sentinel = std::numeric_limits<edge_t>::max(),
-                       d_src    = final_src_v.data().get()] __device__(edge_t edge_id) {
-                        if (edge_id == 0) { return edge_id; }
-
-                        if (d_src[edge_id - 1] != d_src[edge_id]) { return edge_id; }
-
-                        return sentinel;
-                      });
-
-    local_cluster_edge_ids_v = detail::remove_elements_from_vector(
-      local_cluster_edge_ids_v,
-      [sentinel = std::numeric_limits<edge_t>::max()] __device__(auto edge_id) {
-        return (edge_id != sentinel);
+      current_graph_view_,
+      thrust::make_zip_iterator(thrust::make_tuple(old_cluster_sum_v.begin(),
+                                                   d_src_vertex_weights_cache_,
+                                                   cluster_subtract_v.begin(),
+                                                   d_src_cluster_cache_)),
+
+      d_dst_cluster_cache_,
+      thrust::make_counting_iterator<vertex_t>(base_vertex_id_),
+      thrust::make_counting_iterator<vertex_t>(base_vertex_id_ + local_num_vertices_),
+      d_dst_cluster_weights_cache_,
```

Comment:
```copy_v_transform_reduce_key_aggregated_out_nbr``` requires ```map_key_first, map_key_last, map_value_first``` to be pre-shuffled (yes... this might be a bit confusing as relabel doesn't require pre-shuffling, so we may need to make this consistent, but this is the current requirement. ```transform_reduce_by_adj_matrix_row(or col)_key``` returns pre-shuffled key, value pairs, so this helps to avoid unnecessary shuffling.

You need to do something like the following, and pass ```keys.beign(), keys.end(), weights.begin()```

I will submit another PR having fixes in the graph primitives side. After these fixes, I am seeing a failure in thrust::sequence call (https://github.com/rapidsai/cugraph/blob/branch-0.19/cpp/src/experimental/louvain.cuh#L524).

```C++ exception with description: tabulate: failed to synchronize ... ```

```
    if (graph_view_t::is_multi_gpu) {
      rmm::device_uvector<vertex_t> keys(local_num_vertices_, handle_.get_stream());
      thrust::sequence(rmm::exec_policy(stream_)->on(stream_), keys.begin(), keys.end(), base_vertex_id_);
      rmm::device_uvector<weight_t> weights(local_num_vertices_, handle_.get_stream());
      thrust::copy(rmm::exec_policy(stream_)->on(stream_),d_dst_cluster_weights_cache_, d_dst_cluster_weights_cache_ + local_num_vertices_, weights.begin());

      auto const comm_size = handle_.get_comms().get_size();
      rmm::device_uvector<vertex_t> rx_keys(0, handle_.get_stream());
      rmm::device_uvector<weight_t> rx_weights(0, handle_.get_stream());
      auto pair_first = thrust::make_zip_iterator(thrust::make_tuple(keys.begin(), weights.begin()));
      std::forward_as_tuple(std::tie(rx_keys, rx_weights), std::ignore) = groupby_gpuid_and_shuffle_values(handle_.get_comms(), pair_first, pair_first + keys.size(), [key_func = cugraph::experimental::detail::compute_gpu_id_from_vertex_t<vertex_t>{comm_size}]__device__(auto val) { return key_func(thrust::get<0>(val)); }, handle_.get_stream());

      keys = std::move(rx_keys);
      weights = std::move(rx_weights);
    }
```

Reply:
Resolved.